rocks-lib/src/lockfile/mod.rs

```diff
@@ -0,0 +1,199 @@
+use std::io::Write;
+use std::{collections::HashMap, fs::File, io::ErrorKind, path::PathBuf};
+
+use eyre::{bail, Result};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+use crate::lua_package::{LuaPackage, PackageName, PackageVersion};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
```

Comment:
question: if we want this to work with nix, we will need to hash the source content (archive/repo/..
) and the rockspec. 

Is this still planned? It would likely require us to change where we add a rock to the lockfile (to after fetching the sources).

Reply:
Right, that's interesting. I believe we'll want to create a separate hash for this, though, but I'm not sure how the complexity maps over to Nix.
The basic ID is useful because other parts of the application can construct the unique ID just from the name, version and pinning information.

If you'd like to check if a rock exists, you don't want to have to read the entire source content, nor do you want to pull down the rockspec and hash that as well, just to look up the rock in question.

I do think that each rock should also have a `sources_hash` or something similar, which is a hash of the rockspec and the source content. The rockspec already contains version and name information, and pinning information isn't important in a source hash. Wdyt?

Reply:
I don't think we need it for lookups. But it's useful for integrity checks, in case someone force-uploaded a modified rockspec or source archive (which defeats the purpose of locking and is a supply chain vulnerability).
The benefit of this for Nix is that it can use a cached version if the source or rockspec hash has changed. We could one day implement something similar.
And if there's nothing in the cache and a mismatched rockspec/source hash, it would be reported as an error (with an option to update the lock entry, warning that it could be a supply chain attack).

Reply:
Created #116, so we can merge this first.